pkg/azuredisk/nodeserver_v2.go

```diff
@@ -91,25 +133,24 @@ func (d *DriverV2) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolume
 		return nil, status.Errorf(codes.Internal, "Failed to find disk on lun %s. %v", lunStr, err)
 	}
 
-	source, err := d.nodeProvisioner.GetDevicePathWithLUN(int(lun))
-	if err != nil {
-		// check if AzVolumeAttachment reports attached
-		if state, getErr := d.crdProvisioner.GetAzVolumeAttachmentState(ctx, diskURI, d.NodeID); state == nil || getErr != nil {
-			getErr = status.Errorf(codes.Internal, "failed to get current attachment state for volume (%s) and node (%s): %v", diskURI, d.NodeID, getErr)
-			klog.Error(getErr)
-			return nil, getErr
-		} else {
-			if *state != v1alpha1.Attached {
-				getErr = status.Errorf(codes.Internal, "volume (%s) is not yet attached to node (%s)", diskURI, d.NodeID)
-				return nil, getErr
-			}
-		}
-		// when lun is available but device path cannot be found, try recovery by detaching the volume from node
-		// and reattaching it, which is triggered by the difference in world states (listVolume result vs. volumeattachment list)
-		if detachErr := d.crdProvisioner.UnpublishVolume(ctx, diskURI, d.NodeID, nil); detachErr != nil {
-			klog.Errorf("failed to unpublishVolume volume (%s) from node (%s): %v", diskURI, d.NodeID, detachErr)
+	// check if AzVolumeattachment is in Attached state if recovery is in process for the specified volume and is not in process of detaching
+	if _, diskURIMatches := d.isRecoveryInProcess(diskURI, false); diskURIMatches {
+		if state, getErr := d.crdProvisioner.GetAzVolumeAttachmentState(ctx, diskURI, d.NodeID); getErr != nil {
+			return nil, status.Errorf(codes.Internal, "failed to get current attachment state for volume (%s) and node (%s): %v", diskURI, d.NodeID, getErr)
+		} else if d.isDetachInProcess(diskURI) {
+			return nil, status.Errorf(codes.Internal, "recovery for volume (%s) is still in process", diskURI)
+		} else if state != v1alpha1.Attached {
+			return nil, status.Errorf(codes.Internal, "volume (%s) is not yet attached to node (%s)", diskURI, d.NodeID)
 		}
-		return nil, status.Errorf(codes.Internal, "Failed to find disk on lun %v. %v", lun, err)
+	}
+
+	source, err := d.nodeProvisioner.GetDevicePathWithLUN(ctx, int(lun))
+	if err == nil {
+		d.markRecoveryComplete(diskURI)
```

Comment:
The name `markRecoveryComplete` is confusing because we are implementing a logic of doing one recovery per node for possibly many failing volumes. The name indicates that we are marking the recovery of the whole node as complete. i.e. while in the process of recovery, if we get a `/NodeStageVolume` call from a volume that is attached and mounted and `GetDevicePathWithLUN` succeeds for that volume, we mark the recovery complete independent of the result of the actual recovery of failing volumes. The name should be more along the line `markVolumeRecoveryCompleteIfInProcess`

Reply:
Yeah, I can see how the name can be confusing. I will change it:)